include/util/cell_storage.hpp

```diff
@@ -0,0 +1,314 @@
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
+            : num_source_nodes{data.num_source_nodes},
+              num_destination_nodes{data.num_destination_nodes},
+              weights{all_weight + data.weight_offset},
+              source_boundary{all_sources + data.source_boundary_offset},
+              destination_boundary{all_destinations + data.destination_boundary_offset}
+        {
+            BOOST_ASSERT(all_weight != nullptr);
+            BOOST_ASSERT(all_sources != nullptr);
+            BOOST_ASSERT(all_destinations != nullptr);
+        }
+    };
+
+    std::size_t LevelIDToIndex(LevelID level) const { return level - 1; }
+
+  public:
+    using Cell = CellImpl<EdgeWeight>;
+    using ConstCell = CellImpl<const EdgeWeight>;
+
+    template <typename GraphT>
+    CellStorage(const MultiLevelPartition &partition, const GraphT &base_graph)
+    {
+        // pre-allocate storge for CellData so we can have random access to it by cell id
+        unsigned number_of_cells = 0;
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            level_to_cell_offset.push_back(number_of_cells);
+            number_of_cells += partition.GetNumberOfCells(level);
+        }
+        level_to_cell_offset.push_back(number_of_cells);
+        cells.resize(number_of_cells);
+
+        std::vector<std::pair<CellID, NodeID>> level_source_boundary;
+        std::vector<std::pair<CellID, NodeID>> level_destination_boundary;
+
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            auto level_offset = level_to_cell_offset[LevelIDToIndex(level)];
+
+            level_source_boundary.clear();
+            level_destination_boundary.clear();
+
+            for (auto node = 0u; node < base_graph.GetNumberOfNodes(); ++node)
+            {
+                const CellID cell_id = partition.GetCell(level, node);
+                bool is_source_node = false;
+                bool is_destination_node = false;
+                bool is_boundary_node = false;
+
+                for (auto edge : base_graph.GetAdjacentEdgeRange(node))
+                {
+                    auto other = base_graph.GetTarget(edge);
+                    const auto &data = base_graph.GetEdgeData(edge);
+
+                    is_boundary_node |= partition.GetCell(level, other) != cell_id;
+                    is_source_node |= partition.GetCell(level, other) == cell_id && data.forward;
+                    is_destination_node |=
+                        partition.GetCell(level, other) == cell_id && data.backward;
+                }
+
+                if (is_boundary_node)
+                {
+                    if (is_source_node)
+                        level_source_boundary.emplace_back(cell_id, node);
+                    if (is_destination_node)
+                        level_destination_boundary.emplace_back(cell_id, node);
```

Comment:
It is hard to estimate and depends heavily on the quality of the partition. I think 5% of all nodes is a good estimate for level 1, and every level above that will be much below this anyway.